Nodejs/Product/Nodejs/Project/NodeModulesNode.cs

```diff
@@ -378,39 +393,76 @@ private class NpmPathProvider : INpmPathProvider {
             }
 
             var controller = _npmController;
-            if (null != controller) {
-                if (null != RootPackage) {
-                    var dev = controller.RootPackage.Modules.Where(package => package.IsDevDependency);
-                    _devModulesNode.Packages = dev;
-                    ReloadHierarchy(_devModulesNode, dev);
-
-                    var optional = controller.RootPackage.Modules.Where(package => package.IsOptionalDependency);
-                    _optionalModulesNode.Packages = optional;
-                    ReloadHierarchy(_optionalModulesNode, optional);
-
-                    var root = controller.RootPackage.Modules.Where(package => 
-                        package.IsDependency || 
-                        !package.IsListedInParentPackageJson);
-                    
-                    ReloadHierarchy(this, root);
-                }
+            if (null == controller)
+                return;
 
-                var global = controller.GlobalPackages;
-                if (null != global) {
-                    _globalModulesNode.GlobalPackages = global;
-                    ReloadHierarchy(_globalModulesNode, global.Modules);
-                }
+            var newPackages = ReloadPackageHierarchies(controller);
 
-                if (_firstHierarchyLoad) {
-                    controller.FinishedRefresh += NpmController_FinishedRefresh;
-                    _firstHierarchyLoad = false;
-                }
+            if (_firstHierarchyLoad) {
+                controller.FinishedRefresh += NpmController_FinishedRefresh;
+                _firstHierarchyLoad = false;
+            }
+
+            if (EnableAutomaticTypeAcquisition) {
+                TypingAcquisition.AquireTypings(
+                    controller.ListBaseDirectory,
+                    controller.FullPathToRootPackageDirectory,
+                    newPackages.Added,
+                    null).ContinueWith(x => x);
+            }
+        }
+
+        private PackageSet.Diff ReloadPackageHierarchies(INpmController controller) {
+            return ReloadDevPackageHierarchy(controller)
+                .Concat(ReloadOptionalPackageHierarchy(controller))
+                .Concat(ReloadRootPackageHierarchy(controller))
+                .Concat(ReloadGlobalPackageHierarchy(controller));
+        }
```

Comment:
As mentioned earlier, this isn't the right place for this code. Additionally the flow is pretty confusing, and if you choose to keep this refactoring even after moving the code out into `NodejsProjectNode`, please try and refactor them out into using the same method, otherwise the flow becomes annoying to decipher.